src/main/kotlin/org/veupathdb/vdi/lib/common/compression/Zip.kt

```diff
@@ -40,5 +44,76 @@ object Zip {
         ZipType.Invalid
     }
   }
-}
 
+  /**
+   * Unzips the target path into the given [into] directory.
+   *
+   * @receiver Path to the zip file to unzip.
+   *
+   * @param into Directory into which the zip file contents should be unzipped.
+   *
+   * **NOTE** This directory must exist and be empty at the time of this method
+   * call.
+   *
+   * @return A collection of paths to the files that were unzipped.
+   */
+  fun Path.unzip(into: Path): Collection<Path> {
+    if (!into.exists())
+      throw IllegalStateException("cannot unzip $this into non-existent path $into")
+
+    if (!into.isDirectory())
+      throw IllegalStateException("cannot unzip $this into non-directory path $into")
+
+    if (into.listDirectoryEntries().isNotEmpty())
+      throw IllegalStateException("refusing to unzip $this into non-empty directory $into")
+
+    val unzipped = ArrayList<Path>(2)
+
+    zipEntries(this).forEach { (entry, zipStream) ->
+      val target = into.resolve(entry.name)
+
+      if (entry.isDirectory) {
+        target.createDirectories()
+      } else {
+        if (target.parent != into)
+          target.parent.createDirectories()
+
+        target.outputStream().use { outStream -> zipStream.transferTo(outStream) }
+
+        unzipped.add(target)
+      }
+    }
+
+    return unzipped
+  }
+
+  /**
+   * Returns a sequence of [ZipEntry] instances paired with an [InputStream]
+   * that may be used to stream out the contents of the zip entry itself.
+   *
+   * @param zip Path to the zip file whose entries should be sequenced.
+   *
+   * @return A sequence of [ZipEntry] and [InputStream] instances contained in
+   * the target zip file.
+   */
+  fun zipEntries(zip: Path): Sequence<Pair<ZipEntry, InputStream>> = sequence {
+    if (!zip.exists())
+      throw FileNotFoundException("cannot open non-existent zip file $zip")
+
+    if (zip.isDirectory())
+      throw IllegalStateException("cannot open target zip file $zip as it is a directory")
+
+    if (!zip.isReadable())
+      throw IllegalStateException("cannot open unreadable zip file $zip")
+
+    zip.inputStream().use { raw ->
+      val stream = ZipInputStream(raw)
+      var entry  = stream.nextEntry
+
+      while (entry != null) {
+        yield(entry to UncloseableInputStream(stream))
```

Comment:
I'm not sure this is going to work.  Don't you need to create a new stream for each entry?  If you don't then all the streams in the sequence will be the same and pointing to the final entry.  Or is the point of the sequence to consume the streams as they are yielded?  I feel like if that doesn't happen (say they are collected into a List), then this will break.

Reply:
True, if the input streams were collected into a collection rather than consumed immediately, then this would break.  I'm not sure how to avoid that though, because that is true of ZipInputStream in general.

Reply:
To insulate against it I think you'd have to create a new ZipInputStream for each entry, then skip to the entry in question and yield it to the sequence.  Something like:
```
  int entriesToSkip = 0;
  while (true) {
    zip.inputStream().use { raw ->
      val stream = ZipInputStream(raw)
      var entry  = stream.nextEntry
      for (var i = 0; i < entriesToSkip; i++)
        entry = stream.nextEntry
      if (entry == null) break;
      yield(entry to UncloseableInputStream(stream))
      entriesToSkip++;
    }
  }
```

Reply:
You'd have to close them all though.  Probably not worth it, but needs a comment about how the result has to be processed to not break things.